simple_parsing/replace_selections.py

```diff
@@ -0,0 +1,141 @@
+    is_dataclass_type,
+    is_optional,
+    is_union,
+    unflatten_keyword,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@overload
+def replace_selections(
+    obj: DataclassT,
+    changes_dict: dict[str, Any],
+    selections: dict[str, Key | DataclassT] | None = None,
+) -> DataclassT:
+    ...
+
+
+@overload
+def replace_selections(
+    obj: DataclassT, selections: dict[str, Key | DataclassT] | None = None, **changes
+) -> DataclassT:
+    ...
+
+
+def replace_selected_dataclass(
+    obj: DataclassT, selections: dict[str, Key | DataclassT] | None = None
+):
+    """
+    This function replaces the dataclass of subgroups, union, and optional union.
+    The `selections` is in flat format, e.g. {"ab_or_cd": 'cd', "ab_or_cd.c_or_d": 'd'}
+
+    The values of selections can be `Key` of subgroups, dataclass type, and dataclass instance.
+    """
+    keyword = "__key__"
+
+    if selections:
+        selections = unflatten_keyword(selections, keyword)
+    else:
+        return obj
+
+    replace_kwargs = {}
+    for field in dataclasses.fields(obj):
+        if field.name not in selections:
+            continue
+
+        field_value = getattr(obj, field.name)
+        t = get_field_type_from_annotations(obj.__class__, field.name)
+
+        if contains_dataclass_type_arg(t) and is_union(t):
+            child_selections = selections.pop(field.name)
+            key = child_selections.pop(keyword, None)
+
+            if is_dataclass_type(key):
+                field_value = key()
+                logger.debug("is_dataclass_type")
+            elif is_dataclass_instance(key):
+                field_value = copy.deepcopy(key)
+                logger.debug("is_dataclass_instance")
+            elif field.metadata.get("subgroups", None):
+                field_value = field.metadata["subgroups"][key]()
+                logger.debug("is_subgroups")
+            elif is_optional(t) and key is None:
+                field_value = None
+                logger.debug("key is None")
+            else:
+                raise ValueError(f"invalid selection key '{key}' for field '{field.name}'")
+
+            if child_selections:
+                new_value = replace_selected_dataclass(field_value, child_selections)
+            else:
+                new_value = field_value
+
+            if not field.init:
+                raise ValueError(f"Cannot replace value of non-init field {field.name}.")
+
+        replace_kwargs[field.name] = new_value
+    return dataclasses.replace(obj, **replace_kwargs)
+
+
+def replace_selections(
+    obj: DataclassT,
+    changes_dict: dict[str, Any] | None = None,
+    selections: dict[str, Key | DataclassT] | None = None,
+    **changes,
+) -> DataclassT:
+    """Replace some values in a dataclass and replace dataclass type in nested union of dataclasses or subgroups.
+
+    Compared to `simple_replace.replace`, this calls `replace_selected_dataclass` before calling `simple_parsing.replace`.
```

Comment:
Can't we just add `replace_selected_dataclass` then?

I'm having trouble understanding all the use-cases for this function.
Correct me if I'm wrong, but you want to be able to replace:
- Subgroup choices with a key
- Subgroup choices with a subgroup value (dataclass type)
- Subgroup choices with a dataclass instance
- Union fields with a value (makes sense)
- Union fields with a dataclass type (?)

I feel like the API is a bit too complicated with all these arguments.
What about having a `replace_subgroup` function, and letting the user compose the two in their own code, just to keep things simpler?
What do you think?

Reply:
> Subgroup choices with a key
Subgroup choices with a subgroup value (dataclass type)
Subgroup choices with a dataclass instance
Union fields with a value (makes sense)
Union fields with a dataclass type (?)

Yes you are right.
Union fields with dataclass type will be initialized working like a default factory.